### PR TITLE
fix(iast): context rework

### PR DESCRIPTION
**Pull request by avara1986**
his fix ensures that only the IAST propagation context is cleared instead of all contexts,  which could otherwise cause propagation loss in multithreaded applications. Additionally, it improves validations in both the Processor and Vulnerability Reporter, depending on whether IAST is active or not.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Review by juanjux — APPROVED** [review]

File: ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp [hunk 1]
```diff
@@ -198,23 +200,46 @@ Initializer::release_taint_range(TaintRangePtr rangeptr)
 void
 Initializer::create_context()
 {
-    if (ThreadContextCache.tx_map != nullptr) {
-        // Reset the current context
-        reset_context();
+    // const auto log = get_python_logger();
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp:203]:
```suggestion
```

File: ddtrace/appsec/_iast/_taint_tracking/tests/test_common.hpp [hunk 2]
```diff
@@ -1,7 +1,7 @@
 #pragma once
 #include <Initializer/Initializer.h>
-#include <Utils/GenericUtils.h>
 #include <gtest/gtest.h>
+#include <iostream> // JJJ
```
Comment (gnufede) [inline ddtrace/appsec/_iast/_taint_tracking/tests/test_common.hpp:4]:
```suggestion
```

File: ddtrace/appsec/_iast/processor.py [hunk 3]
```diff
@@ -63,12 +59,13 @@ def on_span_finish(self, span: Span):
         if span.span_type != SpanTypes.WEB:
             return
 
+        from ._taint_tracking import reset_context  # noqa: F401
+
         if not core.get_item(IAST.REQUEST_IAST_ENABLED, span=span):
             span.set_metric(IAST.ENABLED, 0.0)
+            reset_context()
```
Comment (gnufede) [inline ddtrace/appsec/_iast/processor.py:66]: Question, do we need this reset context? It's still called on line 81

If we need it regardless, I'd move it to line 63 and remove it from here and 81

File: ddtrace/appsec/_iast/taint_sinks/_base.py [hunk 4]
```diff
@@ -91,6 +92,9 @@ def _prepare_report(cls, span, vulnerability_type, evidence, file_name, line_num
     @classmethod
     def report(cls, evidence_value: Text = "", dialect: Optional[Text] = None) -> None:
         """Build a IastSpanReporter instance to report it in the `AppSecIastSpanProcessor` as a string JSON"""
+        if not _is_iast_enabled():
+            return None
```
Comment (gnufede) [inline ddtrace/appsec/_iast/taint_sinks/_base.py:96]: Do we still need to move this code from the processor/on_span_start to here? Maybe we can do this in a different PR? I'd rather have them separate in case we need to revert

File: ddtrace/appsec/_iast/processor.py (hunk 3 above)
Comment (avara1986) [reply ddtrace/appsec/_iast/processor.py:66]: No, I can't move it. we need in line 66 because a early return. reset_context must called  at 81 because we still need context in `build_and_scrub_value_parts`, `_set_metric_iast_request_tainted` and `_set_span_tag_iast_executed_sink`

File: releasenotes/notes/iast-fix-context-rework-3b0f1b54ab4f4156.yaml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+fixes:
+  - |
+    Security (IAST): This fix ensures that only the IAST propagation context is cleared instead of all contexts,
```
Comment (gnufede) [inline releasenotes/notes/iast-fix-context-rework-3b0f1b54ab4f4156.yaml:4]:
```suggestion
    Code Security (IAST): This fix ensures that only the IAST propagation context is cleared instead of all contexts, 
```

**Review by gnufede — APPROVED** [review]

File: ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp [hunk 6]
```diff
@@ -1,10 +1,13 @@
 #include "Initializer.h"
 
+#include <mutex>
 #include <thread>
 
 using namespace std;
 using namespace pybind11::literals;
 
+std::mutex tx_map_mutex;
```
Comment (juanjux) [inline ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp]: This would be better as a member of Initalizer I think. Also, should the name be active_map_addreses_mutex? The tx_map is ThreadLocal so there is no sharing with it.

File: ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp [hunk 7]
```diff
@@ -34,39 +37,43 @@ Initializer::create_tainting_map()
...
     if (const auto it = active_map_addreses.find(tx_map.get()); it == active_map_addreses.end()) {
-        // Map wasn't in the active addresses, do nothing
         return;
     }
 
     for (const auto& [fst, snd] : *tx_map) {
         snd.second->decref();
     }
-
     tx_map->clear();
```
Comment (juanjux) [inline ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp:47]: mutex here? Or at the start of the function (after lowering the one in clear_tainting_maps, see below)

File: ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp [hunk 8]
```diff
@@ -34,39 +37,43 @@ Initializer::create_tainting_map()
...
-    return ThreadContextCache.tx_map;
+    active_map_addreses.erase(tx_map.get());
 }
 
 void
 Initializer::clear_tainting_maps()
 {
     // Need to copy because free_tainting_map changes the set inside the iteration
-    for (auto& [fst, snd] : initializer->active_map_addreses) {
+    std::lock_guard<std::mutex> lock(tx_map_mutex);
```
Comment (juanjux) [inline ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp:55]: mutex here not needed if we add it to clear_tainting_map

File: ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp [hunk 9]
```diff
@@ -34,39 +37,43 @@ Initializer::create_tainting_map()
...
+    auto copy_active_map_addreses(initializer->active_map_addreses);
+    for (auto& [fst, snd] : copy_active_map_addreses) {
+        if (copy_active_map_addreses.empty()) {
+            break;
+        }
         clear_tainting_map(snd);
         snd = nullptr;
     }
     active_map_addreses.clear();
```
Comment (juanjux) [inline ddtrace/appsec/_iast/_taint_tracking/Initializer/Initializer.cpp:64]: mutex above this if we remove the above

**Review by juanjux — APPROVED** [review]
